### PR TITLE
[codegen/schema] Add package name allowlist

### LINKED ISSUE #8556: Add allowlist of non-package names in Pulumi schema
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently, we enforce that all type tokens defined in the schema are prefixed under the package name. This prevents certain use cases (CRD wrappers). We should add a field to the schema that lists additional valid prefixs. We should then only error on unexpected prefixs. 

- We need to be careful to ensure test coverage for tokens with other names. 

### Affected area/feature

schema

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

Fixes #8556

## Checklist

- [X] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/schema/schema.go [hunk 1]
```diff
@@ -1831,12 +1834,15 @@ const (
 )
 
 // Validate an individual name token.
-func (spec *PackageSpec) validateTypeToken(section, token string) hcl.Diagnostics {
+func (spec *PackageSpec) validateTypeToken(validPrefixes map[string]bool, section, token string) hcl.Diagnostics {
```
Comment (justinvp) [inline pkg/codegen/schema/schema.go:1837]: Mind using the same naming throughout for improved readability/understandability?

```suggestion
func (spec *PackageSpec) validateTypeToken(allowedPackageNames map[string]bool, section, token string) hcl.Diagnostics {
```

File: pkg/codegen/schema/schema.go [hunk 2]
```diff
@@ -1831,12 +1834,15 @@ const (
...
 	diags := hcl.Diagnostics{}
 
-	requiredPrefix := spec.Name + ":"
 	path := memberPath(section, token)
-	if !strings.HasPrefix(token, requiredPrefix) {
+	var prefix string
+	if i := strings.Index(token, ":"); i != -1 {
+		prefix = token[:i]
+	}
+	if !validPrefixes[prefix] {
```
Comment (justinvp) [inline pkg/codegen/schema/schema.go:1845]:
```suggestion
	var packageName string
	if i := strings.Index(token, ":"); i != -1 {
		packageName = token[:i]
	}
	if !allowedPackageNames[packageName] {
```

File: pkg/codegen/schema/schema.go [hunk 3]
```diff
@@ -1847,14 +1853,18 @@ func (spec *PackageSpec) validateTypeToken(section, token string) hcl.Diagnostic
 // This is for validating non-reference type tokens.
 func (spec *PackageSpec) validateTypeTokens() hcl.Diagnostics {
 	diags := hcl.Diagnostics{}
+	validPrefixes := map[string]bool{spec.Name: true}
+	for _, prefix := range spec.AllowedPackageNames {
+		validPrefixes[prefix] = true
+	}
```
Comment (justinvp) [inline pkg/codegen/schema/schema.go:1859]:
```suggestion
	allowedPackageNames := map[string]bool{spec.Name: true}
	for _, packageName := range spec.AllowedPackageNames {
		allowedPackageNames[packageName] = true
	}
```

**Review by justinvp — APPROVED** [review]
LGTM

File: pkg/codegen/schema/schema.go (hunk 1 above)
Comment (iwahbe) [reply pkg/codegen/schema/schema.go]: Good point

**Pull request merged**